Topics/Tech_Stacks/Web-Sockets.md

```diff
@@ -0,0 +1,162 @@
+
+### How the WebSocket Connection is Established:
+1. **Client Handshake Request:** The client initiates the connection by sending a WebSocket handshake request to the server. This request looks like a standard HTTP request but includes an `Upgrade: websocket` header that signals the desire to establish a WebSocket connection.
+   
+2. **Server Handshake Response:** If the server supports WebSockets and accepts the connection request, it responds with a handshake of its own, confirming the upgrade from HTTP to WebSockets.
+   
+3. **Persistent Connection:** With the handshake complete, the connection remains open, and data can now be sent in either direction at any time, allowing for two-way, **real-time communication**.
+
+### WebSocket Communication:
+Once established, the WebSocket connection allows messages to be passed back and forth between the client and server with minimal overhead. Each message is framed in a standardized format, which can be quickly parsed and processed by both parties. This efficient messaging system is key to the protocol's ability to support high-frequency, real-time communication scenarios.
+
+WebSockets are particularly adept at handling use cases where the server needs to send data to the client proactively, without the client first making a request. This capability has led to the widespread adoption of WebSockets in various applications that require real-time functionality, such as live notifications, instant messaging, and collaborative editing platforms.
+
+In the context of CSC301, understanding WebSockets is essential for building modern applications where real-time user interaction is a core requirement. With WebSockets, software engineers can create responsive and interactive user experiences that were once difficult to achieve with traditional web technologies.
+
+## WebSockets vs HTTP
+While HTTP is a foundational web technology, its request-response model can introduce latency and inefficiency, particularly in scenarios requiring real-time updates. WebSockets address these limitations by establishing a persistent, full-duplex communication channel, offering several advantages:
+
+- **Reduced Latency:** WebSockets eliminate the need for repeated HTTP handshakes, significantly decreasing latency.
+- **Real-time Data Flow:** The persistent connection allows for real-time data exchange, critical for applications such as online gaming, live sports updates, and financial trading platforms.
+- **Efficient Resource Utilization:** With a single connection per client, WebSockets use server and network resources more efficiently than HTTP polling.
+
+## Advantages and Disadvantages
+## Advantages and Disadvantages
+
+### Advantages
```

Comment:
Could also talk about how WebSockets are beneficial for different types of server, suppose if client is using large server or small servers.

Reply:
Noted.